Data/HashMap/Strict.hs

```diff
@@ -249,6 +259,112 @@ alter f k m =
     Just v  -> insert k v m
 {-# INLINABLE alter #-}
 
+-- | /O(log n)/  The expression (@'alterF' f k map@) alters the value @x@ at
+-- @k@, or absence thereof. @alterF@ can be used to insert, delete, or update
+-- a value in a map.
+--
+-- Note: 'alterF' is a flipped version of the 'at' combinator from
+-- <https://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-At.html#v:at Control.Lens.At>.
+--
+-- @since 0.2.9
+alterF :: (Functor f, Eq k, Hashable k)
+       => (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+-- Special care is taken to only calculate the hash and only perform a key
+-- comparison once.
+alterF f !k !m = (<$> f mv) $ \fres ->
+  case fres of
+    Nothing -> delete' h k m
+    Just !v' -> insert' h k v' m
+
+  where !h = hash k
+        mv = lookup' h k m
+{-# INLINABLE [0] alterF #-}
+
+#if MIN_VERSION_base(4,8,0)
+bogus :: a
+bogus = error "Data.HashMap.alterF internal error: hit bottom"
+
+bogus# :: (# #) -> (# a #)
+bogus# _ = error "Data.HashMap.alterF internal error: hit bottom#"
+
+{-# RULES
+
+-- See detailed notes on alterF rules in Data.HashMap.Base.
+
+"alterFWeird" forall f k m. alterF f k m = alterFWeird (f Nothing) (f (Just bogus)) f k m
+
+"alterFinsert" forall (f :: Maybe a -> Identity (Maybe a)) x k m.
+  alterFWeird (coerce (Just x)) (coerce (Just x)) f k m =
+    Identity (insert k x m)
+
+"alterFinsertWith" [1] forall (f :: Maybe a -> Identity (Maybe a)) x y k m.
+  alterFWeird (coerce (Just x)) (coerce (Just y)) f k m =
+    Identity (insertModifying x (\mold -> case runIdentity (f (Just mold)) of
+                                            Nothing -> bogus# (# #)
+                                            Just !new -> (# new #)) k m)
+
+-- This rule is written a bit differently than the one for lazy
+-- maps because the adjust here is strict. We could write it the
+-- same general way anyway, but this seems simpler.
+"alterFadjust" forall (f :: Maybe a -> Identity (Maybe a)) x k m.
+  alterFWeird (coerce Nothing) (coerce (Just x)) f k m =
+    Identity (adjust (\a -> case runIdentity (f (Just a)) of
+                               Just a' -> a'
+                               Nothing -> bogus) k m)
+
+"alterFdelete" forall (f :: Maybe a -> Identity (Maybe a)) k m.
+  alterFWeird (coerce Nothing) (coerce Nothing) f k m = Identity (delete k m)
+
+"alterFlookup" forall _ign1 _ign2 (f :: Maybe a -> Const r (Maybe a)) k m.
+  alterFWeird _ign1 _ign2 f k m = k `seq` m `seq` Const (getConst (f (lookup k m)))
```

Comment:
Out of curiosity, can you not use bang patterns in rewrite rules?

Reply:
I doubt you can use them on the left-hand side.

Reply:
Yeah, I didn't think so.